ignite/metrics/mean_average_precision.py

```diff
@@ -0,0 +1,440 @@
+        if crowd is not None:
+            iscrowd = crowd[g]
+        else:
+            iscrowd = 0
+        for d in range(m):
+            y_pred_bbox = y_pred[d].tolist()
+            y_pred_area = y_pred_bbox[2] * y_pred_bbox[3]
+            ious[d, g] = 0
+            w = min(y_pred_bbox[2] + y_pred_bbox[0], y_bbox[2] + y_bbox[0]) - max(y_pred_bbox[0], y_bbox[0])
+            h = min(y_pred_bbox[3] + y_pred_bbox[1], y_bbox[3] + y_bbox[1]) - max(y_pred_bbox[1], y_bbox[1])
+            if w <= 0 or h <= 0:
+                continue
+            intersection = w * h
+            if iscrowd:
+                union = y_pred_area
+            else:
+                union = y_pred_area + y_area - intersection
+            ious[d, g] = intersection / union
+
+    return ious
+
+
+class MeanAveragePrecision(Metric):
+    r"""Calculates Mean Average Precision (mAP) for object detection data.
+
+    .. math::
+       \text{MeanAveragePrecision} = {1/11}\sum_{r \in {0.0, ..., 1.0}}\text{AP}_r
+
+    where \text{MeanAveragePrecision}_r represents average precision at recall value :math:`r`,
+    where :math:`r` ranges from 0.0 to 1.0.
+
+    More details can be found in `Everingham et al. 2009`__
+
+    __ https://homepages.inf.ed.ac.uk/ckiw/postscript/ijcv_voc09.pdf
+
+    Remark:
+
+        This implementation is inspired by pycocotools which can be found `here`__
+
+        __ https://github.com/cocodataset/cocoapi/
+
+    Args:
+        object_area_ranges: dictionary with area_name as key and a tuple of the form
+            (lower_area_range, upper_area_range) as value, where both values are floats.
+            It contains all area ranges for which AP needs to be computed.
+        num_detection_max: Number of maximum detections allowed per image. This is to limit crowding and repeated
+            detections of the same object.
+        iou_thresholds: Optional list of float IoU thresholds for which AP is computed (default: [.5:.05:.95]).
+        rec_thresholds: Optional list of float values to which AP is computed for averaging (default: [0:.01:1]).
+        output_transform: a callable that is used to transform the
+            :class:`~ignite.engine.engine.Engine`'s ``process_function``'s output into the
+            form expected by the metric. This can be useful if, for example, you have a multi-output model and
+            you want to compute the metric with respect to one of the outputs.
+            By default, metrics require the output as ``(y_pred, y)`` or ``{'y_pred': y_pred, 'y': y}``.
+        device: specifies which device updates are accumulated on. Setting the
+            metric's device to be the same as your ``update`` arguments ensures the ``update`` method is
+            non-blocking. By default, CPU.
+
+    Example:
+
+        .. code-block:: python
+
+            import torch
+            from ignite.metrics import MeanAveragePrecision
+
+            # Input Format:
+            # Ground Truth
+            # [
```

Comment:
Sizes of tensor (ie B) have no sense. We talked about a list of dict, one per image. So image_id should be an int (or a tensor of size one), etc.